// @nrwl/vite/src/executors/preview-server/preview-server.impl.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    vitePreviewServerExecutor: ()=>vitePreviewServerExecutor,
    default: ()=>_default
});
const _extends = require("@swc/helpers/lib/_extends.js").default;
const _devkit = require("@nrwl/devkit");
const _vite = require("vite");
const _optionsUtils = require("../../utils/options-utils");
async function* vitePreviewServerExecutor(options, context) {
    var ref;
    const target = (0, _devkit.parseTargetString)(options.buildTarget, context.projectGraph);
    const targetConfiguration = (ref = context.projectsConfigurations.projects[target.project]) == null ? void 0 : ref.targets[target.target];
    if (!targetConfiguration) {
        throw new Error(`Invalid buildTarget: ${options.buildTarget}`);
    }
    const isCustomBuildTarget = targetConfiguration.executor !== '@nrwl/vite:build';
    // Retrieve the option for the configured buildTarget.
    const buildTargetOptions = (0, _optionsUtils.getNxTargetOptions)(options.buildTarget, context);
    var _staticFilePath;
    const outputPath = (_staticFilePath = options.staticFilePath) != null ? _staticFilePath : buildTargetOptions.outputPath;
    if (!outputPath) {
        throw new Error(`Could not infer the "outputPath". It should either be a property of the "${options.buildTarget}" buildTarget or provided explicitly as a "staticFilePath" option.`);
    }
    // Merge the options from the build and preview-serve targets.
    // The latter takes precedence.
    const mergedOptions = _extends({}, {
        watch: {}
    }, isCustomBuildTarget ? {} : buildTargetOptions, options, {
        outputPath
    });
    // Retrieve the server configuration.
    const serverConfig = (0, _vite.mergeConfig)((0, _optionsUtils.getViteSharedConfig)(mergedOptions, options.clearScreen, context), {
        build: (0, _optionsUtils.getViteBuildOptions)(mergedOptions, context),
        preview: (0, _optionsUtils.getVitePreviewOptions)(mergedOptions, context)
    });
    if (serverConfig.mode === 'production') {
        console.warn('WARNING: preview is not meant to be run in production!');
    }
    let server;
    const processOnExit = async ()=>{
        await closeServer(server);
    };
    process.once('SIGINT', processOnExit);
    process.once('SIGTERM', processOnExit);
    process.once('exit', processOnExit);
    // Launch the build target.
    // If customBuildTarget is set to true, do not provide any overrides to it
    const buildTargetOverrides = isCustomBuildTarget ? {} : mergedOptions;
    const build = await (0, _devkit.runExecutor)(target, buildTargetOverrides, context);
    for await (const result of build){
        if (result.success) {
            try {
                if (!server) {
                    server = await (0, _vite.preview)(serverConfig);
                }
                server.printUrls();
                const resolvedUrls = [
                    ...server.resolvedUrls.local,
                    ...server.resolvedUrls.network
                ];
                var ref1;
                yield {
                    success: true,
                    baseUrl: (ref1 = resolvedUrls[0]) != null ? ref1 : ''
                };
            } catch (e) {
                console.error(e);
                yield {
                    success: false,
                    baseUrl: ''
                };
            }
        } else {
            yield {
                success: false,
                baseUrl: ''
            };
        }
    }
    await new Promise((resolve)=>{
        process.once('SIGINT', ()=>resolve());
        process.once('SIGTERM', ()=>resolve());
        process.once('exit', ()=>resolve());
    });
}
function closeServer(server) {
    return new Promise((resolve)=>{
        if (!server) {
            resolve();
        } else {
            const { httpServer  } = server;
            // closeAllConnections was added in Node v18.2.0
            httpServer.closeAllConnections && httpServer.closeAllConnections();
            httpServer.close(()=>resolve());
        }
    });
}
const _default = vitePreviewServerExecutor;
