// @nrwl/vite/src/utils/versions.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    nxVersion: ()=>nxVersion,
    viteVersion: ()=>viteVersion,
    vitePluginEslintVersion: ()=>vitePluginEslintVersion,
    vitestVersion: ()=>vitestVersion,
    vitestUiVersion: ()=>vitestUiVersion,
    vitePluginReactVersion: ()=>vitePluginReactVersion,
    vitePluginVueVersion: ()=>vitePluginVueVersion,
    vitePluginVueJsxVersion: ()=>vitePluginVueJsxVersion,
    viteTsConfigPathsVersion: ()=>viteTsConfigPathsVersion,
    jsdomVersion: ()=>jsdomVersion,
    vitePluginDtsVersion: ()=>vitePluginDtsVersion,
    vitestCoverageC8Version: ()=>vitestCoverageC8Version,
    vitestCoverageIstanbulVersion: ()=>vitestCoverageIstanbulVersion
});
const nxVersion = require('../../package.json').version;
const viteVersion = '^4.0.1';
const vitePluginEslintVersion = '^1.8.1';
const vitestVersion = '^0.25.8';
const vitestUiVersion = '^0.25.8';
const vitePluginReactVersion = '^3.0.0';
const vitePluginVueVersion = '^3.2.0';
const vitePluginVueJsxVersion = '^2.1.1';
const viteTsConfigPathsVersion = '^4.0.2';
const jsdomVersion = '~20.0.3';
const vitePluginDtsVersion = '~1.7.1';
const vitestCoverageC8Version = '~0.25.8';
const vitestCoverageIstanbulVersion = '~0.25.8';
