// @nrwl/react/plugins/storybook/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webpack = exports.core = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const config_1 = require("@nrwl/webpack/src/utils/config");
const path_1 = require("path");
const webpack_1 = require("webpack");
const merge_plugins_1 = require("./merge-plugins");
const with_react_1 = require("../with-react");
// This is shamelessly taken from CRA and modified for NX use
// https://github.com/facebook/create-react-app/blob/4784997f0682e75eb32a897b4ffe34d735912e6c/packages/react-scripts/config/env.js#L71
function getClientEnvironment(mode) {
    // Grab NODE_ENV and NX_* and STORYBOOK_* environment variables and prepare them to be
    // injected into the application via DefinePlugin in webpack configuration.
    const NX_PREFIX = /^NX_/i;
    const STORYBOOK_PREFIX = /^STORYBOOK_/i;
    const raw = Object.keys(process.env)
        .filter((key) => NX_PREFIX.test(key) || STORYBOOK_PREFIX.test(key))
        .reduce((env, key) => {
        env[key] = process.env[key];
        return env;
    }, {
        // Useful for determining whether we’re running in production mode.
        NODE_ENV: process.env.NODE_ENV || mode,
        // Environment variables for Storybook
        // https://github.com/storybookjs/storybook/blob/bdf9e5ed854b8d34e737eee1a4a05add88265e92/lib/core-common/src/utils/envs.ts#L12-L21
        NODE_PATH: process.env.NODE_PATH || '',
        STORYBOOK: process.env.STORYBOOK || 'true',
        // This is to support CRA's public folder feature.
        // In production we set this to dot(.) to allow the browser to access these assets
        // even when deployed inside a subpath. (like in GitHub pages)
        // In development this is just empty as we always serves from the root.
        PUBLIC_URL: mode === 'production' ? '.' : '',
    });
    // Stringify all values so we can feed into webpack DefinePlugin
    const stringified = {
        'process.env': Object.keys(raw).reduce((env, key) => {
            env[key] = JSON.stringify(raw[key]);
            return env;
        }, {}),
    };
    return { stringified };
}
const core = (prev, options) => (Object.assign(Object.assign({}, prev), { disableWebpackDefaults: true }));
exports.core = core;
const webpack = (storybookWebpackConfig = {}, options) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e;
    devkit_1.logger.info('=> Loading Nx React Storybook Addon from "@nrwl/react/plugins/storybook"');
    // In case anyone is missing dep and did not run migrations.
    // See: https://github.com/nrwl/nx/issues/14455
    try {
        require.resolve('@nrwl/webpack');
    }
    catch (_f) {
        throw new Error(`'@nrwl/webpack' package is not installed. Install it and try again.`);
    }
    const { withNx, withWeb } = require('@nrwl/webpack');
    const tsconfigPath = (0, path_1.join)(options.configDir, 'tsconfig.json');
    const builderOptions = Object.assign(Object.assign({}, options), { root: options.configDir, 
        // These are blank because root is the absolute path to .storybook folder
        projectRoot: '', sourceRoot: '', fileReplacements: [], sourceMap: true, styles: (_a = options.styles) !== null && _a !== void 0 ? _a : [], optimization: {}, tsConfig: tsconfigPath, extractCss: storybookWebpackConfig.mode === 'production', target: 'web' });
    // ESM build for modern browsers.
    let baseWebpackConfig = {};
    const configure = (0, config_1.composePluginsSync)(withNx({ skipTypeChecking: true }), withWeb(), (0, with_react_1.withReact)());
    const finalConfig = configure(baseWebpackConfig, {
        options: builderOptions,
        context: { root: devkit_1.workspaceRoot }, // The context is not used here.
    });
    return Object.assign(Object.assign({}, storybookWebpackConfig), { module: Object.assign(Object.assign({}, storybookWebpackConfig.module), { rules: [
                ...storybookWebpackConfig.module.rules,
                ...finalConfig.module.rules,
            ] }), resolve: Object.assign(Object.assign({}, storybookWebpackConfig.resolve), { plugins: (0, merge_plugins_1.mergePlugins)(...((_b = storybookWebpackConfig.resolve.plugins) !== null && _b !== void 0 ? _b : []), ...((_c = finalConfig.resolve
                .plugins) !== null && _c !== void 0 ? _c : [])) }), plugins: (0, merge_plugins_1.mergePlugins)(new webpack_1.DefinePlugin(getClientEnvironment(storybookWebpackConfig.mode).stringified), ...((_d = storybookWebpackConfig.plugins) !== null && _d !== void 0 ? _d : []), ...((_e = finalConfig.plugins) !== null && _e !== void 0 ? _e : [])) });
});
exports.webpack = webpack;
