// @nrwl/nx-cloud/lib/utilities/nx-imports.js
"use strict";
try {
    try {
        const { output } = require('nx/src/utils/output');
        let workspaceRoot;
        try {
            workspaceRoot = require('nx/src/utils/app-root').workspaceRoot;
        }
        catch (_a) {
            workspaceRoot = require('nx/src/utils/workspace-root').workspaceRoot;
        }
        const { getDependencyConfigs } = require('nx/src/tasks-runner/utils');
        const tasksRunner = require('nx/tasks-runners/default').default;
        const { CompositeLifeCycle } = require('nx/src/tasks-runner/life-cycle');
        let initTasksRunner = null;
        try {
            initTasksRunner = require('nx/src/index').initTasksRunner;
        }
        catch (e) { }
        exports.runnerReturnsPromise = true;
        exports.output = output;
        exports.workspaceRoot = workspaceRoot;
        exports.tasksRunner = tasksRunner;
        exports.CompositeLifeCycle = CompositeLifeCycle;
        exports.getDependencyConfigs = getDependencyConfigs;
        exports.initTasksRunner = initTasksRunner;
    }
    catch (e) {
        const { output } = require('@nrwl/workspace/src/utilities/output');
        const { appRootPath } = require('@nrwl/tao/src/utils/app-root');
        const { getDependencyConfigs, } = require('@nrwl/workspace/src/tasks-runner/utils');
        const { tasksRunnerV2, } = require('@nrwl/workspace/src/tasks-runner/tasks-runner-v2');
        let CompositeLifeCycle;
        try {
            CompositeLifeCycle =
                require('@nrwl/workspace/src/tasks-runner/life-cycle').CompositeLifeCycle;
        }
        catch (e) { }
        exports.runnerReturnsPromise = false;
        exports.output = output;
        exports.workspaceRoot = appRootPath;
        exports.tasksRunner = tasksRunnerV2;
        exports.CompositeLifeCycle = CompositeLifeCycle;
        exports.getDependencyConfigs = getDependencyConfigs;
        exports.initTasksRunner = null;
    }
}
catch (e) {
    if (process.env.NX_VERBOSE_LOGGING === 'true') {
        console.log(e);
    }
    console.error('NX CLOUD ERROR');
    console.error('---------------------------------------');
    console.error('This version of Nx Cloud is incompatible with the @nrwl/* packages in your workspace.');
    console.error('');
    console.error('Match your @nrwl/nx-cloud version to the same major version of your @nrwl/* packages and try again.');
    console.error('---------------------------------------');
    process.exit(1);
}
