// @nrwl/react/src/generators/storybook-configuration/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storybookConfigurationSchematic = exports.storybookConfigurationGenerator = void 0;
const tslib_1 = require("tslib");
const stories_1 = require("../stories/stories");
const devkit_1 = require("@nrwl/devkit");
const versions_1 = require("../../utils/versions");
function generateStories(host, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        (0, devkit_1.ensurePackage)('@nrwl/cypress', versions_1.nxVersion);
        const { getE2eProjectName } = yield Promise.resolve().then(() => require('@nrwl/cypress/src/utils/project-name'));
        const projectConfig = (0, devkit_1.readProjectConfiguration)(host, schema.name);
        const cypressProject = getE2eProjectName(schema.name, projectConfig.root, schema.cypressDirectory);
        yield (0, stories_1.default)(host, {
            project: schema.name,
            generateCypressSpecs: schema.configureCypress && schema.generateCypressSpecs,
            js: schema.js,
            cypressProject,
            ignorePaths: schema.ignorePaths,
        });
    });
}
function storybookConfigurationGenerator(host, schema) {
    var _a, _b;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { configurationGenerator } = (0, devkit_1.ensurePackage)('@nrwl/storybook', versions_1.nxVersion);
        let bundler = (_a = schema.bundler) !== null && _a !== void 0 ? _a : 'webpack';
        const projectConfig = (0, devkit_1.readProjectConfiguration)(host, schema.name);
        if (projectConfig.projectType === 'application' &&
            ((_b = projectConfig.targets['build']) === null || _b === void 0 ? void 0 : _b.executor) === '@nrwl/vite:build') {
            bundler = 'vite';
            if (schema.bundler !== 'vite') {
                devkit_1.logger.info(`The project ${schema.name} is set up to use Vite. So
      Storybook will be configured to use Vite as well.`);
            }
        }
        const installTask = yield configurationGenerator(host, {
            name: schema.name,
            uiFramework: '@storybook/react',
            configureCypress: schema.configureCypress,
            js: schema.js,
            linter: schema.linter,
            cypressDirectory: schema.cypressDirectory,
            standaloneConfig: schema.standaloneConfig,
            tsConfiguration: schema.tsConfiguration,
            configureTestRunner: schema.configureTestRunner,
            bundler,
            storybook7betaConfiguration: schema.storybook7betaConfiguration,
            storybook7UiFramework: bundler === 'vite'
                ? '@storybook/react-vite'
                : '@storybook/react-webpack5',
        });
        if (schema.generateStories) {
            yield generateStories(host, schema);
        }
        return installTask;
    });
}
exports.storybookConfigurationGenerator = storybookConfigurationGenerator;
exports.default = storybookConfigurationGenerator;
exports.storybookConfigurationSchematic = (0, devkit_1.convertNxGenerator)(storybookConfigurationGenerator);
